book-content/chapters/05-unions-literals-and-narrowing.md

```diff
@@ -0,0 +1,1856 @@
+
+#### Exercise 3: Using `in` to Narrow
+
+Here we have a `handleResponse` function that takes in a type of `APIResponse`, which is a union of two types of objects.
+
+The goal of the `handleResponse` function is to check whether the provided object has a `data` property. If it does, the function should return the `id` property. If not, it should throw an `Error` with the message from the `error` property.
+
+```tsx
+type APIResponse =
+  | {
+      data: {
+        id: string;
+      };
+    }
+  | {
+      error: string;
+    };
+
+const handleResponse = (response: APIResponse) => {
+  // How do we check if 'data' is in the response?
+
+  if (true) {
+    return response.data.id;
+  } else {
+    throw new Error(response.error);
+  }
+};
+```
+
+Currently, there are several errors being thrown as seen in the following tests.
+
+The first error is `Property 'data' does not exist on type 'APIResponse'`
+
+```tsx
+test("passes the test even with the error", () => {
+  // there is no error in runtime
+
+  expect(() =>
+    HandleResponseOrThrowError({
+      error: "Invalid argument",
+    }),
+  ).not.toThrowError();
+
+  // but the data is returned, instead of an error.
+
+  expect(
+    HandleResponseOrThrowError({
+      error: "Invalid argument",
+    }),
+  ).toEqual("Should this be 'Error'?");
+});
+```
+
+Then we have the inverse error, where `Property 'error' does not exist on type 'APIResponse'`:
+
+```
+Property data does not exist on type 'APIResponse'.
+```
+
+Your challenge is to find the correct syntax for narrowing down the types within the `handleResponse` function's `if` condition.
+
+The changes should happen inside of the function without modifying any other parts of the code.
+
+<Exercise title="Exercise 3: Using `in` to Narrow" filePath="/src/018-unions-and-narrowing/064-narrowing-with-in-statements.problem.ts"></Exercise>
+
+#### Solution 1: Narrowing with `if` Statements
+
+There are several different ways to validate the username length.
+
+##### Option 1: Check Truthiness
+
+We could use an `if` statement to check if `username` is truthy. If it does, we can return `username.length > 5`, otherwise we can return `false`:
+
+```typescript
+function validateUsername(username: string | null): boolean {
+  // Rewrite this function to make the error go away
+
+  if (username) {
+    return username.length > 5;
+  }
+
+  return false;
+}
+```
+
+There is a catch to this piece of logic. If `username` is an empty string, it will return `false` because an empty string is falsy. This happens to match the behavior we want for this exercise - but it's important to bear that in mind.
+
+##### Option 2: Check if `typeof username` is `"string"`
+
+We could use `typeof` to check if the username is a string:
+
+```typescript
+function validateUsername(username: string | null): boolean {
+  if (typeof username === "string") {
+    return username.length > 5;
+  }
+
+  return false;
+}
+```
+
+This avoids the issue with empty strings.
```

Comment:
it sounds like it avoids it *here* but it avoids it in other cases for which it might matter, it's irrelevant here (which u called out above but perhaps this statement here could still be more precise)